src/main/scala/io/iohk/ethereum/db/dataSource/RocksDbDataSource.scala

```diff
@@ -101,6 +106,43 @@ class RocksDbDataSource(
     }
   }
 
+  private def dbIterator: Resource[Task, RocksIterator] = {
+    Resource.fromAutoCloseable(Task(db.newIterator()))
+  }
+
+  private def namespaceIterator(namespace: Namespace): Resource[Task, RocksIterator] = {
+    Resource.fromAutoCloseable(Task(db.newIterator(handles(namespace))))
+  }
+
+  private def moveIterator(it: RocksIterator): Observable[Either[IterationError, (Array[Byte], Array[Byte])]] = {
+    Observable
+      .fromTask(Task(it.seekToFirst()))
+      .flatMap(_ => Observable.fromTask(Task(it.isValid)))
+      .filter(identity)
+      .flatMap { _ =>
+        Observable.fromTask(Task(Right(it.key(), it.value()))) ++ Observable
+          .repeatEvalF {
+            for {
+              _ <- Task(it.next())
+              isValid <- Task(it.isValid)
+              result <- if (isValid) Task(Right(it.key(), it.value())) else Task.raiseError(IterationFinished)
+            } yield result
+          }
+          .onErrorHandleWith {
+            case IterationFinished => Observable.empty
+            case ex => Observable(Left(IterationError(ex)))
+          }
+      }
+  }
+
+  def iterate(): Observable[Either[IterationError, (Array[Byte], Array[Byte])]] = {
+    Observable.fromResource(dbIterator).flatMap(it => moveIterator(it))
```

Comment:
what about calling `it.seekToFirst` here?

Reply:
I do not think that spreading iterator movements across two function is good idea. Imo it is nice to see it and have it in one place